proposals/p5434.md

```diff
@@ -0,0 +1,1331 @@
+
+We saw how much we were leaning on value bindings. The asymmetry between having
+value binding but not referenced bindings when have value expressions and
+reference expressions was creating pressure. For example, when accessing members
+of an object, we had to escape to pointers in that operator.
+
+One downside of this change is that before indirections were more visible in the
+code.
+
+Also, this does fundamentally mean that we now have another kind of "pointer",
+potentially adding complexity to any memory-safety story. However, this ship
+already sailed to some extent with value bindings. Fundamentally, bindings are
+allowed to have pointer-like semantics from a lifetime perspective, and so will
+need to be considered as a pointer-like thing as we build out lifetime safety.
+
+### Remove pointers after adding references
+
+If we removed pointers after adding references, we would need something
+rebindable for assignable objects. The viable path forward without separate
+pointers and references is to have something rebindable like pointers but
+automatically dereferenced like references, which is the approach Rust takes.
+See
+[this comment on issue #5261](https://github.com/carbon-language/carbon-lang/issues/5261#issuecomment-2786462775).
+
+One of the features of a reference is what it cannot do, and we would have to
+remove those restrictions to be able to satisfy the pointer use cases with
+references.
+
+### Allow `ref` bindings in the fields of classes
+
+A type with reference binding fields would need a lot of restrictions since
+reference bindings are not assignable. We did not see enough motivation to put
+references into objects, given the complexity that it would introduce, so we are
+keeping references out of types for now. This could change to support lambda
+reference captures.
+
+### No call-site annotation
+
+This question was discussed on
+[2025-05-07](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.sfx9d7ltud5).
+
+We decided that the marking is not about lifetime, but ability to mutate. A
+`let` may reference an object in a similar way to a `ref`, restricting
+operations on the original object, but we are not going to mark `let`s since
+those restrictions are enforced by the compiler. We thought the ability to
+mutate, though, was something important enough to highlight to readers of the
+code, even at the expense of extra work for the writer.
+
+Swift `inout` parameters are
+[marked at caller with an `&` before the argument](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/functions/#In-Out-Parameters).
+Jordon Rose has published
+[a regret](https://belkadan.com/blog/2021/12/Swift-Regret-inout-Syntax/) that
+they didn't use `inout` to mark the argument instead of `&`.
+
+On the other hand, not marking is not known to be a source of bugs.
+
+This is a "try it and see how well it works" sort of decision.
+
+### Top-level `ref` introducer
```

Comment:
Just got finished reading! Really excited about this direction.

One question comes to mind:

> Top-level ref introducer
> 
> For now, we don't believe let ref to be so common as to need a shorter way to write, unlike what we do for var.

There doesn't appear to be much written on this. Why was this decision made? I personally find it surprising that the following code is invalid:

```
fn F(ptr: T*) {
  ref t: T = *ptr;
}
```

Is this something that we have strong consensus on? Or is it possible to reconsider before landing the proposal?

Thanks!

Reply:
(Marking as resolved, this seems to have a good body of rationale now.)